contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,652 @@
+  )
+    private
+    returns (
+      bool success
+    )
+  {
+    assembly{
+      let g := gas()
+      // Compute g -= CUSHION and check for underflow
+      if lt(g, CUSHION) { revert(0, 0) }
+      g := sub(g, CUSHION)
+      // if g - g//64 <= gasAmount, revert
+      // (we subtract g//64 because of EIP-150)
+      if iszero(gt(sub(g, div(g, 64)), gasAmount)) { revert(0, 0) }
+      // solidity calls check that a contract actually exists at the destination, so we do the same
+      if iszero(extcodesize(target)) { revert(0, 0) }
+      // call and return whether we succeeded. ignore return data
+      success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
+    }
+    return success;
+  }
+
+  // Offsets into fulfillRandomnessRequest's proof of various values
+  //
+  // Public key. Skips byte array's length prefix.
+  uint256 public constant PUBLIC_KEY_OFFSET = 0x20;
+  // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
+  uint256 public constant PRESEED_OFFSET = 7*0x20;
+
+  function fulfillRandomWords(
+    bytes memory proof
+  )
+    external
+  {
+    uint256 startGas = gasleft();
+    (bytes32 keyHash, uint256 requestId,
+    uint256 randomness, FulfillmentParams memory fp) = getRandomnessFromProof(proof);
+
+
+    uint256[] memory randomWords = new uint256[](fp.numWords);
+    for (uint256 i = 0; i < fp.numWords; i++) {
+      randomWords[i] = uint256(keccak256(abi.encode(randomness, i)));
+    }
+
+    // Prevent re-entrancy. The user callback cannot call fulfillRandomWords again
+    // with the same proof because this getRandomnessFromProof will revert because the requestId
+    // is gone.
+    delete s_callbacks[requestId];
+    VRFConsumerV2Interface v;
+    bytes memory resp = abi.encodeWithSelector(v.fulfillRandomWords.selector, requestId, randomWords);
+    uint256 gasPreCallback = gasleft();
+    if (gasPreCallback < fp.callbackGasLimit) {
+      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
+    }
+    // Call with explicitly the amount of callback gas requested
+    // Important to not let them exhaust the gas budget and avoid oracle payment.
+    bool success = callWithExactGas(fp.callbackGasLimit, fp.sender, resp);
+    emit RandomWordsFulfilled(requestId, randomWords, success);
+
+    // We want to charge users exactly for how much gas they use in their callback.
+    // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+    // decrement the subscription balance and increment the oracles withdrawable balance.
+    // We also add the flat link fee to the payment amount.
+    // Its specified in millionths of link, if s_config.fulfillmentFlatFeeLinkPPM = 1
+    // 1 link / 1e6 = 1e18 juels / 1e6 = 1e12 juels.
+    uint96 payment = calculatePaymentAmount(startGas, s_config.gasAfterPaymentCalculation, s_config.fulfillmentFlatFeeLinkPPM, tx.gasprice);
+    if (s_subscriptions[fp.subId].balance < payment) {
```

Comment:
Same possible optimisation for `s_subscriptions` here.

Reply:
Similar reasons to in the request flow, want to avoid a potentially large sload with all the consumers